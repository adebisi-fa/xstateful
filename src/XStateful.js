// @flow

import Emitter from 'tiny-emitter';
import { StateNode, actions } from 'xstate';

import type { ExtState, State, Event, EventObject } from './types';

type Constructor = {|
    machine: StateNode,
    extstate: ExtState,
|};

class XStateful extends Emitter {
    machine: StateNode;

    state: State;
    extstate: ExtState;

    pvt: {
        ctChangeGuard: number,
        initialExtState: ExtState,
        initialState: State,
    };

    constructor({ machine, extstate }: Constructor) {
        super();

        this.machine = machine;
        this.extstate = extstate;
        this.state = null;

        this.pvt = {
            ctChangeGuard: 0,
            initialExtState: extstate,
            initialState: machine.initialState,
        };
    }

    init() {
        this.state = this.pvt.initialState;
        this.extstate = this.pvt.initialExtState;

        // need to handle any onEntry actions for the initial state.
        // these actions may modify extstate, causing a change notification.
        // as we want to ensure only one change notification, we add a guard.
        this.pvtGuardChangeNotification(() => {
            this.pvtInvokeActions(null);
        });
    }

    transition(event: Event) {
        const statePrevious = this.state;
        const stateNext = this.machine.transition(
            this.state,
            event,
            this.extstate,
        );

        if (stateNext === statePrevious) {
            return;
        }

        this.pvtGuardChangeNotification(() => {
            this.state = stateNext;
            this.pvtInvokeActions(actions.toEventObject(event)); // might call setExtState
        });
    }

    setExtState(updater: ?ExtState | ((xs: ExtState) => ?ExtState)) {
        const update =
            typeof updater === 'function' ? updater(this.extstate) : updater;

        if (!update) {
            return;
        }

        this.pvtGuardChangeNotification(() => {
            this.extstate = { ...this.extstate, ...update };
        });
    }

    // avoid multiple change notifications caused by recursive calls to setExtState via action events.
    pvtGuardChangeNotification(fn: () => void) {
        this.pvt.ctChangeGuard += 1;

        fn();

        this.pvt.ctChangeGuard -= 1;

        if (this.pvt.ctChangeGuard === 0) {
            this.emit('change', { state: this.state, extstate: this.extstate });
        }
    }

    pvtInvokeActions(event: EventObject) {
        if (this.state.actions.length === 0) {
            return;
        }

        const actionObjs = this.state.actions.map(action =>
            actions.toActionObject(action),
        );

        this.emit('before-actions', actionObjs.slice());

        actionObjs.map(action =>
            this.emit('action', {
                state: this.state,
                extstate: this.extstate,
                event,
                action: actions.toActionObject(action),
            }),
        );

        this.emit('after-actions', actionObjs.slice());
    }
}

export default XStateful;
