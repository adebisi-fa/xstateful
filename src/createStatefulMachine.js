// @flow

import { StateNode } from 'xstate';

import type { CreateStatefulMachine } from './types';

import XStateful from './XStateful';
import Reducer from './Reducer';
import getReducerResult from './getReducerResult';
import applyReducerResult from './applyReducerResult';

const argDefault = {
    machine: null,
    reducer: Reducer.map({}),
    extstate: {},
};

const createStatefulMachine: CreateStatefulMachine = arg => {
    const { machine, reducer, extstate } = {
        ...argDefault,
        ...arg,
    };

    if (!(machine instanceof StateNode)) {
        throw new Error(
            'You must pass an instantiated xstate machine to createStatefulMachine',
        );
    }

    const xsf = new XStateful({ machine, extstate });
    let effects = [];

    xsf.on('before-actions', () => {
        effects = [];
    });

    xsf.on('action', reducerarg => {
        const effect = applyReducerResult({
            xsf,
            reducerarg,
            reducerresult: getReducerResult({ reducerarg, reducer }),
        });

        if (effect) {
            effects.push(effect);
        }
    });

    xsf.on('after-actions', () => {
        effects.forEach(eff => eff());
    });

    return xsf;
};

export default createStatefulMachine;
