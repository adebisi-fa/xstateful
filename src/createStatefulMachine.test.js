// @flow

import type { MachineConfig } from 'xstate';
import { Machine } from 'xstate';

import createStatefulMachine from './createStatefulMachine';
import XStateful from './XStateful';

const config: MachineConfig = {
    key: 'updown',
    initial: 'down',
    states: {
        up: {
            on: {
                SWITCH: 'down',
            },
        },
        down: {
            on: {
                SWITCH: 'up',
            },
        },
    },
};

describe('createStatefulMachine', () => {
    test('cannot pass nothing', () => {
        const fn = () => {
            // $FlowFixMe (deliberate error)
            createStatefulMachine();
        };

        expect(fn).toThrowErrorMatchingSnapshot();
    });

    test('must pass a machine key', () => {
        const fn = () => {
            // $FlowFixMe (deliberate error)
            createStatefulMachine({});
        };

        expect(fn).toThrowErrorMatchingSnapshot();
    });

    test('cannot pass an xstate config as machine', () => {
        const fn = () => {
            // $FlowFixMe (deliberate error)
            createStatefulMachine({ machine: config });
        };

        expect(fn).toThrowErrorMatchingSnapshot();
    });

    test('on success, return is an instance of XStateful', () => {
        const machine = Machine(config);
        const xsf = createStatefulMachine({ machine });

        expect(xsf).toBeInstanceOf(XStateful);
    });

    test('can pass extstate through to XStateful instance', () => {
        const machine = Machine(config);
        const extstate = { foo: 123 };
        const xsf = createStatefulMachine({ machine, extstate });

        expect(xsf.extstate).toBe(extstate);
    });

    test('can pass reducer which ends up seeing actions', () => {
        const machine = Machine({
            key: 'test',
            initial: 'first',
            states: {
                first: {
                    onEntry: 'ping',
                },
            },
        });
        const reducer = jest.fn();

        // $FlowFixMe (deliberate error)
        const xsf = createStatefulMachine({ machine, reducer });
        xsf.init();

        expect(reducer).toHaveBeenCalledTimes(1);
        expect(reducer.mock.calls[0]).toHaveLength(1);
        expect(reducer.mock.calls[0][0]).toHaveProperty('action');
        expect(reducer.mock.calls[0][0].action.type).toBe('ping');
    });
});
