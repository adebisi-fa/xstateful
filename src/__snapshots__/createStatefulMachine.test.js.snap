// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createStatefulMachine cannot pass an xstate config as machine 1`] = `"You must pass an instantiated xstate machine to createStatefulMachine"`;

exports[`createStatefulMachine cannot pass nothing 1`] = `"You must pass an instantiated xstate machine to createStatefulMachine"`;

exports[`createStatefulMachine must pass a machine key 1`] = `"You must pass an instantiated xstate machine to createStatefulMachine"`;
